[DEV APPROVED] Redirect to 404

shaley91 opened the pull request:
If the requested page number is more than the total number of pages we redirect them to a 404 page.

There was a bug where the search index page would display no matter what page number you requested.  So if there were 100 results, and 10 results per page, there should only be 10 pages.  However if you updated the page param to be 11 you still got a page that would link to page 10, and page 12.  This was what was causing a lot of extra pages to be indexed by search engines.

norm commented:
Correct HTTP behaviour would be to *return* a status code of 404, not a 301/302 redirect to a page with a 200 status code.

Tr4pSt3R commented on app/controllers/search_controller.rb:
```diff
@@ -5,6 +5,7 @@ def index
     @form = SearchForm.new(search_form_params)
     if @form.valid?
       @result = FirmRepository.new.search(@form.to_query, page: page)
+      redirect_to "/404.html" if @result.total_pages < page
```
@shaley91 could you use single-quotes in this instance please?

"/404.html" --> '/404.html'

pmcostello replied:
simple fix 

render :file => 'public/404.html',  :status => :not_found, :layout => false if @result.total_pages < page

shaley91 replied:
@pmcostello does your simple fix update the url?  Surely we want to actually redirect the page so that it doesn't get indexed?

Tr4pSt3R replied:
@shaley91 why don't you create a non-RESTful action which you can redirect to?

norm replied:
If the page returns a 404 status code, that will stop Google indexing it, because you are declaring "this isn't a valid page" rather than "what you're looking for is actually over there".

pmcostello replied:
what Norm said ☝️ the 404 status is the important bit to stop the crawl errors and it will return the 404.html page

shaley91 replied:
Awesome.  I learnt something.

margOnline commented on spec/controllers/search_controller_spec.rb:
```diff
@@ -0,0 +1,25 @@
+      end
+    end
+  end
+end
```
I like that you've added tests for a controller that didn't **but obviously should have had tests**. Please would you add tests for the other paths of the index method.

pmcostello commented:
there are two problems on this ticket, 
- generating a 404 status and loading the 404 page, 
&
- forward and back links appearing on pages when the requested page number is more than the total number of pages

the second issue I think can be fixed by updating the 'kaminari' gem used for pagination as the "Hide Next & Last buttons if page is out of range" bug is fixed in version 1.0.0

https://github.com/kaminari/kaminari/blob/master/CHANGELOG.md

shaley91 commented:
@pmcostello we have updated the redirect to just render the 404 page.  This now means that the issue with the forward and back links being display is now fixed as you cannot get to that page.

Tr4pSt3R commented:
@shaley91 is it worth updating the `kaminari` gem from _0.16.3_ to _1.0.0_? As suggested by @pmcostello? Mind you it has a lot of breaking changes. If this current patch is enough then we can leave it be.

`Hide Next & Last buttons if page is out of range #712 [@igorkasyanchuk]`

shaley91 commented:
If it has a lot of breaking changes then I really don't think it is sensible to upgrade.  The upgrade is also not required any more to fix the issue.

pmcostello commented:
thanks @shaley91 looks good, that should fix it,
@Tr4pSt3R 
i've updated the kaminari gem locally and all appears to work well and tests pass ok, but agree with Stephen it's not required to fix the specific behaviour on this ticket

norm commented:
Could we reword the commit message and spec description to not mention "redirecting"?
